File: TarsEngine/Services/CSharpStructureExtractor.cs
using System.Text.RegularExpressions;
using Microsoft.Extensions.Logging;
using TarsEngine.Models;
using TarsEngine.Services.Interfaces;

namespace TarsEngine.Services;

/// <summary>
/// Extracts code structures from C# code
/// </summary>
public class CSharpStructureExtractor(ILogger<CSharpStructureExtractor> logger) : ICodeStructureExtractor
{
    private readonly ILogger<CSharpStructureExtractor> _logger = logger;

    /// <inheritdoc/>
    public string Language => "csharp";

    /// <inheritdoc/>
    public List<CodeStructure> ExtractStructures(string content)
    {
        var structures = new List<CodeStructure>();

        try
        {
            if (string.IsNullOrWhiteSpace(content))
            {
                return structures;
            }

            // Extract namespaces
            var namespaceRegex = new Regex(@"namespace\s+([a-zA-Z0-9_.]+)\s*{", RegexOptions.Compiled);
            var namespaceMatches = namespaceRegex.Matches(content);
            foreach (Match match in namespaceMatches)
            {
                if (match.Groups.Count > 1)
                {
                    var namespaceName = match.Groups[1].Value;
                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Namespace,
                        Name = namespaceName,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = namespaceName
                        }
                    });
                }
            }

            // Extract classes
            var classRegex = new Regex(@"(public|private|protected|internal)?\s*(static|abstract|sealed)?\s*class\s+([a-zA-Z0-9_]+)(?:<[^>]+>)?(?:\s*:\s*[^{]+)?", RegexOptions.Compiled);
            var classMatches = classRegex.Matches(content);
            foreach (Match match in classMatches)
            {
                if (match.Groups.Count > 3)
                {
                    var className = match.Groups[3].Value;
                    var namespaceName = GetNamespaceForPosition(structures, match.Index, content);
                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Class,
                        Name = className,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = namespaceName,
                            ClassName = className
                        }
                    });
                }
            }

            // Extract interfaces
            var interfaceRegex = new Regex(@"(public|private|protected|internal)?\s*interface\s+([a-zA-Z0-9_]+)(?:<[^>]+>)?(?:\s*:\s*[^{]+)?", RegexOptions.Compiled);
            var interfaceMatches = interfaceRegex.Matches(content);
            foreach (Match match in interfaceMatches)
            {
                if (match.Groups.Count > 2)
                {
                    var interfaceName = match.Groups[2].Value;
                    var namespaceName = GetNamespaceForPosition(structures, match.Index, content);
                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Interface,
                        Name = interfaceName,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = namespaceName,
                            ClassName = interfaceName
                        }
                    });
                }
            }

            // Extract methods
            var methodRegex = new Regex(@"(public|private|protected|internal)?\s*(static|virtual|abstract|override|async)?\s*([a-zA-Z0-9_<>]+)\s+([a-zA-Z0-9_]+)\s*\(", RegexOptions.Compiled);
            var methodMatches = methodRegex.Matches(content);
            foreach (Match match in methodMatches)
            {
                if (match.Groups.Count > 4)
                {
                    var methodName = match.Groups[4].Value;
                    var returnType = match.Groups[3].Value;
                    var className = GetClassForPosition(structures, match.Index, content);
                    var namespaceName = GetNamespaceForPosition(structures, match.Index, content);
                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Method,
                        Name = methodName,
                        ParentName = className,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = namespaceName,
                            ClassName = className
                        },
                        Properties = new Dictionary<string, string>
                        {
                            { "ReturnType", returnType }
                        }
                    });
                }
            }

            // Extract properties
            var propertyRegex = new Regex(@"(public|private|protected|internal)?\s*(static|virtual|abstract|override)?\s*([a-zA-Z0-9_<>]+)\s+([a-zA-Z0-9_]+)\s*{\s*(get|set)?", RegexOptions.Compiled);
            var propertyMatches = propertyRegex.Matches(content);
            foreach (Match match in propertyMatches)
            {
                if (match.Groups.Count > 4)
                {
                    var propertyName = match.Groups[4].Value;
                    var propertyType = match.Groups[3].Value;
                    var className = GetClassForPosition(structures, match.Index, content);
                    var namespaceName = GetNamespaceForPosition(structures, match.Index, content);
                    structures.Add(new CodeStructure
                    {
                        Type = StructureType.Property,
                        Name = propertyName,
                        ParentName = className,
                        Location = new CodeLocation
                        {
                            StartLine = GetLineNumber(content, match.Index),
                            Namespace = namespaceName,
                            ClassName = className
                        },
                        Properties = new Dictionary<string, string>
                        {
                            { "PropertyType", propertyType }
                        }
                    });
                }
            }

            // Calculate structure sizes and update end lines
            CalculateStructureSizes(structures, content);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error extracting structures from C# code");
        }

        return structures;
    }

    /// <inheritdoc/>
    public string GetNamespaceForPosition(List<CodeStructure> structures, int position, string content)
    {
        var lineNumber = GetLineNumber(content, position);
        var namespaceStructures = structures
            .Where(s => s.Type == StructureType.Namespace)
            .ToList();

        foreach (var ns in namespaceStructures)
        {
            if (ns.Location.StartLine <= lineNumber && 
                (ns.Location.EndLine == 0 || ns.Location.EndLine >= lineNumber))
            {
                return ns.Name;
            }
        }

        return string.Empty;
    }

    /// <inheritdoc/>
    public string GetClassForPosition(List<CodeStructure> structures, int position, string content)
    {
        var lineNumber = GetLineNumber(content, position);
        var classStructures = structures
            .Where(s => s.Type == StructureType.Class)
            .ToList();

        foreach (var cls in classStructures)
        {
            if (cls.Location.StartLine <= lineNumber && 
                (cls.Location.EndLine == 0 || cls.Location.EndLine >= lineNumber))
            {
                return cls.Name;
            }
        }

        return string.Empty;
    }

    /// <inheritdoc/>
    public void CalculateStructureSizes(List<CodeStructure> structures, string content)
    {
        // Find matching braces to determine end positions
        foreach (var structure in structures)
        {
            if (structure.Type == StructureType.Namespace || 
                structure.Type == StructureType.Class || 
                structure.Type == StructureType.Interface)
            {
                // Find the opening brace position
                var startLine = structure.Location.StartLine;
                var startLinePos = GetPositionForLine(content, startLine);
                var openBracePos = content.IndexOf('{', startLinePos);
                    
                if (openBracePos != -1)
                {
                    // Find the matching closing brace
                    var endPos = FindMatchingBrace(content, openBracePos);
                    if (endPos != -1)
                    {
                        var endLine = GetLineNumber(content, endPos);
                        structure.Location.EndLine = endLine;
                        structure.Size = endLine - startLine + 1;
                    }
                }
            }
            else if (structure.Type == StructureType.Method)
            {
                // Find the opening brace position
                var startLine = structure.Location.StartLine;
                var startLinePos = GetPositionForLine(content, startLine);
                var openParenPos = content.IndexOf('(', startLinePos);
                    
                if (openParenPos != -1)
                {
                    var closeParenPos = FindMatchingParenthesis(content, openParenPos);
                    if (closeParenPos != -1)
                    {
                        var openBracePos = content.IndexOf('{', closeParenPos);
                        if (openBracePos != -1)
                        {
                            // Find the matching closing brace
                            var endPos = FindMatchingBrace(content, openBracePos);
                            if (endPos != -1)
                            {
                                var endLine = GetLineNumber(content, endPos);
                                structure.Location.EndLine = endLine;
                                structure.Size = endLine - startLine + 1;
                            }
                        }
                    }
                }
            }
            else if (structure.Type == StructureType.Property)
            {
                // Find the opening brace position
                var startLine = structure.Location.StartLine;
                var startLinePos = GetPositionForLine(content, startLine);
                var openBracePos = content.IndexOf('{', startLinePos);
                    
                if (openBracePos != -1)
                {
                    // Find the matching closing brace
                    var endPos = FindMatchingBrace(content, openBracePos);
                    if (endPos != -1)
                    {
                        var endLine = GetLineNumber(content, endPos);
                        structure.Location.EndLine = endLine;
                        structure.Size = endLine - startLine + 1;
                    }
                }
            }
        }
    }

    /// <summary>
    /// Gets the line number for a position in the content
    /// </summary>
    public int GetLineNumber(string content, int position)
    {
        if (string.IsNullOrEmpty(content) || position < 0 || position >= content.Length)
        {
            return 0;
        }

        // Count newlines before the position
        return content[..position].Count(c => c == '\n') + 1;
    }

    /// <summary>
    /// Gets the position for a line number in the content
    /// </summary>
    private int GetPositionForLine(string content, int lineNumber)
    {
        if (string.IsNullOrEmpty(content) || lineNumber <= 0)
        {
            return 0;
        }

        int currentLine = 1;
        int position = 0;

        while (currentLine < lineNumber && position < content.Length)
        {
            if (content[position] == '\n')
            {
                currentLine++;
            }
            position++;
        }

        return position;
    }

    /// <summary>
    /// Finds the position of the matching closing brace
    /// </summary>
    private static int FindMatchingBrace(string content, int openBracePos)
    {
        int braceCount = 1;
        for (int i = openBracePos + 1; i < content.Length; i++)
        {
            if (content[i] == '{')
            {
                braceCount++;
            }
            else if (content[i] == '}')
            {
                braceCount--;
                if (braceCount == 0)
                {
                    return i;
                }
            }
        }
        return -1;
    }

    /// <summary>
    /// Finds the position of the matching closing parenthesis
    /// </summary>
    private static int FindMatchingParenthesis(string content, int openParenPos)
    {
        int parenCount = 1;
        for (int i = openParenPos + 1; i < content.Length; i++)
        {
            if (content[i] == '(')
            {
                parenCount++;
            }
            else if (content[i] == ')')
            {
                parenCount--;
                if (parenCount == 0)
                {
                    return i;
                }
            }
        }
        return -1;
    }
}

File: TarsEngine.FSharp.Core/Analysis/CSharpStructureExtractor.fs
namespace TarsEngine.FSharp.Core.Analysis

open System
open System.Collections.Generic
open System.IO
open System.Text.RegularExpressions
open Microsoft.Extensions.Logging

/// <summary>
/// Implementation of ICodeStructureExtractor for C# language.
/// </summary>
type CSharpStructureExtractor(logger: ILogger<CSharpStructureExtractor>) =
    
    /// <summary>
    /// Gets the language supported by this extractor.
    /// </summary>
    member _.Language = "csharp"
    
    /// <summary>
    /// Extracts code structures from the provided content.
    /// </summary>
    /// <param name="content">The source code content.</param>
    /// <returns>A list of extracted code structures.</returns>
    member _.ExtractStructures(content: string) =
        try
            logger.LogInformation("Extracting structures from C# code")
            
            // Define regex patterns for different structure types
            let namespacePattern = @"namespace\s+([a-zA-Z0-9_\.]+)\s*\{"
            let classPattern = @"(public|private|protected|internal)?\s*(static|abstract|sealed)?\s*class\s+([a-zA-Z0-9_]+)(?:<[^>]+>)?\s*(?::\s*[^{]+)?\s*\{"
            let interfacePattern = @"(public|private|protected|internal)?\s*interface\s+([a-zA-Z0-9_]+)(?:<[^>]+>)?\s*(?::\s*[^{]+)?\s*\{"
            let enumPattern = @"(public|private|protected|internal)?\s*enum\s+([a-zA-Z0-9_]+)\s*\{"
            let structPattern = @"(public|private|protected|internal)?\s*struct\s+([a-zA-Z0-9_]+)(?:<[^>]+>)?\s*(?::\s*[^{]+)?\s*\{"
            let methodPattern = @"(public|private|protected|internal)?\s*(static|virtual|abstract|override|sealed)?\s*([a-zA-Z0-9_<>]+)\s+([a-zA-Z0-9_]+)\s*\(([^)]*)\)\s*(?:where\s+[^{]+)?\s*\{"
            let propertyPattern = @"(public|private|protected|internal)?\s*(static|virtual|abstract|override|sealed)?\s*([a-zA-Z0-9_<>]+)\s+([a-zA-Z0-9_]+)\s*\{\s*(?:get;)?\s*(?:set;)?\s*\}"
            let fieldPattern = @"(public|private|protected|internal)?\s*(static|readonly|const)?\s*([a-zA-Z0-9_<>]+)\s+([a-zA-Z0-9_]+)\s*=?\s*[^;]*;"
            let delegatePattern = @"(public|private|protected|internal)?\s*delegate\s+([a-zA-Z0-9_<>]+)\s+([a-zA-Z0-9_]+)\s*\(([^)]*)\)\s*;"
            let eventPattern = @"(public|private|protected|internal)?\s*event\s+([a-zA-Z0-9_<>]+)\s+([a-zA-Z0-9_]+)\s*;"
            let usingPattern = @"using\s+(?:static\s+)?([a-zA-Z0-9_\.]+)\s*;"
            let attributePattern = @"\[([a-zA-Z0-9_]+)(?:\(.*\))?\]"
            
            // Extract structures using regex
            let extractStructures pattern structureType =
                Regex.Matches(content, pattern, RegexOptions.Multiline)
                |> Seq.cast<Match>
                |> Seq.map (fun m -> 
                    let startLine = content.Substring(0, m.Index).Split('\n').Length
                    let endLine = startLine + m.Value.Split('\n').Length - 1
                    let modifiers = 
                        m.Groups
                        |> Seq.cast<Group>
                        |> Seq.skip 1 // Skip the full match
                        |> Seq.takeWhile (fun g -> g.Success && g.Value <> "")
                        |> Seq.map (fun g -> g.Value)
                        |> Seq.toList
                    
                    let name = 
                        match structureType with
                        | "namespace" -> m.Groups.[1].Value
                        | "class" | "interface" | "enum" | "struct" -> m.Groups.[3].Value
                        | "method" | "property" | "field" | "delegate" | "event" -> m.Groups.[4].Value
                        | "using" | "attribute" -> m.Groups.[1].Value
                        | _ -> ""
                    
                    let returnType = 
                        match structureType with
                        | "method" | "property" | "field" -> Some m.Groups.[3].Value
                        | "delegate" -> Some m.Groups.[2].Value
                        | _ -> None
                    
                    let parameters = 
                        match structureType with
                        | "method" | "delegate" ->
                            let paramString = m.Groups.[5].Value
                            if String.IsNullOrWhiteSpace(paramString) then
                                []
                            else
                                paramString.Split(',')
                                |> Array.map (fun p -> 
                                    let parts = p.Trim().Split(' ')
                                    if parts.Length >= 2 then
                                        (parts.[0], parts.[1])
                                    else
                                        ("", p.Trim())
                                )
                                |> Array.toList
                        | _ -> []
                    
                    {
                        Name = name
                        StructureType = structureType
                        Parent = None
                        Children = []
                        StartLine = startLine
                        EndLine = endLine
                        Modifiers = modifiers
                        ReturnType = returnType
                        Parameters = parameters
                        Properties = Map.empty
                    }
                )
                |> Seq.toList
            
            // Extract all structures
            let namespaces = extractStructures namespacePattern "namespace"
            let classes = extractStructures classPattern "class"
            let interfaces = extractStructures interfacePattern "interface"
            let enums = extractStructures enumPattern "enum"
            let structs = extractStructures structPattern "struct"
            let methods = extractStructures methodPattern "method"
            let properties = extractStructures propertyPattern "property"
            let fields = extractStructures fieldPattern "field"
            let delegates = extractStructures delegatePattern "delegate"
            let events = extractStructures eventPattern "event"
            let usings = extractStructures usingPattern "using"
            let attributes = extractStructures attributePattern "attribute"
            
            // Combine all structures
            let allStructures = 
                List.concat [
                    namespaces
                    classes
                    interfaces
                    enums
                    structs
                    methods
                    properties
                    fields
                    delegates
                    events
                    usings
                    attributes
                ]
            
            // Build parent-child relationships
            let structureMap = Dictionary<string, CodeStructure>()
            
            // Add structures to map
            for structure in allStructures do
                let key = $"{structure.StructureType}:{structure.Name}"
                if not (structureMap.ContainsKey(key)) then
                    structureMap.Add(key, structure)
            
            // Build parent-child relationships
            let structures = 
                allStructures
                |> List.map (fun structure ->
                    // Find parent
                    let parent = 
                        allStructures
                        |> List.filter (fun s -> 
                            s.StartLine < structure.StartLine && 
                            s.EndLine > structure.EndLine && 
                            s.StructureType <> "using" && 
                            s.StructureType <> "attribute"
                        )
                        |> List.sortByDescending (fun s -> s.StartLine)
                        |> List.tryHead
                    
                    // Find children
                    let children = 
                        allStructures
                        |> List.filter (fun s -> 
                            s.StartLine > structure.StartLine && 
                            s.EndLine < structure.EndLine && 
                            s.StructureType <> "using" && 
                            s.StructureType <> "attribute"
                        )
                        |> List.filter (fun s ->
                            // Ensure the child is a direct child
                            not (allStructures |> List.exists (fun p -> 
                                p.StartLine < s.StartLine && 
                                p.EndLine > s.EndLine && 
                                p.StartLine > structure.StartLine && 
                                p.EndLine < structure.EndLine
                            ))
                        )
                    
                    // Update structure with parent and children
                    { structure with 
                        Parent = parent
                        Children = children
                    }
                )
            
            structures
        with
        | ex ->
            logger.LogError(ex, "Error extracting structures from C# code")
            []
    
    /// <summary>
    /// Extracts code structures from a file.
    /// </summary>
    /// <param name="filePath">The path to the file.</param>
    /// <returns>A list of extracted code structures.</returns>
    member this.ExtractStructuresFromFile(filePath: string) =
        try
            logger.LogInformation("Extracting structures from C# file: {FilePath}", filePath)
            
            // Read the file content
            let content = File.ReadAllText(filePath)
            
            // Extract structures from the content
            this.ExtractStructures(content)
        with
        | ex ->
            logger.LogError(ex, "Error extracting structures from C# file: {FilePath}", filePath)
            []
    
    /// <summary>
    /// Gets a structure by name.
    /// </summary>
    /// <param name="structures">The list of structures to search.</param>
    /// <param name="name">The name of the structure to find.</param>
    /// <returns>The found structure, if any.</returns>
    member _.GetStructureByName(structures: CodeStructure list, name: string) =
        structures |> List.tryFind (fun s -> s.Name = name)
    
    /// <summary>
    /// Gets structures by type.
    /// </summary>
    /// <param name="structures">The list of structures to search.</param>
    /// <param name="structureType">The type of structures to find.</param>
    /// <returns>The list of found structures.</returns>
    member _.GetStructuresByType(structures: CodeStructure list, structureType: string) =
        structures |> List.filter (fun s -> s.StructureType = structureType)
    
    interface ICodeStructureExtractor with
        member this.Language = this.Language
        member this.ExtractStructures(content) = this.ExtractStructures(content)
        member this.ExtractStructuresFromFile(filePath) = this.ExtractStructuresFromFile(filePath)
        member this.GetStructureByName(structures, name) = this.GetStructureByName(structures, name)
        member this.GetStructuresByType(structures, structureType) = this.GetStructuresByType(structures, structureType)
